BrainPortal/app/controllers/nh_loris_hooks_controller.rb

```diff
@@ -64,21 +64,70 @@ def file_list_maker
     s_group = Group.where_id_or_name(source_group_id).first if source_group_id
     base = Userfile.where(nil) # As seen in userfiles_controller
     base = Userfile.restrict_access_on_query(current_user, base, :access_requested => :read)
+
+    # Special situation when a file with partial path is specified
+    # instead of just a basename.
+    extra_params_by_parent = {}
+
+    source_basenames = source_basenames.map do |filepath|
+      # file path can either be:
+      #    sub-123
+      #    sub-123/anat/file_T1w.nii.gz
+      filenames =  Pathname.new(filepath).each_filename.to_a
+      # E.g: root == sub-123
+      parent_dir = filenames.first
+      next parent_dir if filenames.size == 1
+
+      # E.g: sub-123/anat/...T1w.nii.gz
+      # Initialisation of the array of filenames to keep
+      current_params = extra_params_by_parent[parent_dir] ||= { :basenames => [] }
+      # E.g: basename == file_T1w.nii.gz
+      basename = filenames.last
+      next parent_dir if current_params[:basenames].include?(basename)
+      current_params[:basenames] << basename
+
+      parent_dir
+    end.uniq
+
     userfiles = base.where(:name => source_basenames)
     userfiles = userfiles.where(:data_provider_id => s_dp.id)    if s_dp
     userfiles = userfiles.where(:group_id         => s_group.id) if s_group
 
+    # Add setter and getter for json_params in Userfile class
+    Userfile.class_eval do
```

Comment:
You shouldn't do that. You're modifying the Userfile class for ALL classes and subclasses.

What you want is defined object methods JUSt on the objects that are you pass along to the ExtendedCbrainFileList constructor.

Reply:
Use singleton methods

Reply:
The creation of the method shoudl be performed by a utility method in the ExtendedCbrainFileList class:

```ruby
  # in ExtendedCbrainFileList
  def self.extend_userfile_list(userfiles, extra_info = {}) ; extra info is { basename_or_uid? => RUBYTHING }
     #add singleton method :json_params to return RUBYTHING
  end